wrappers/hnyecho/echo.go

```diff
@@ -0,0 +1,81 @@
+package hnyecho
+
+import (
+	"github.com/jamietsao/beeline-go/wrappers/common"
+	"github.com/labstack/echo"
+)
+
+// EchoWrapper provides Honeycomb instrumentation for the Echo router via middleware
+type (
+	EchoWrapper struct {
+		handlerNames map[string]string
+	}
+)
+
+// New returns a new EchoWrapper struct
+func New() *EchoWrapper {
+	return &EchoWrapper{}
+}
+
+// Middleware returns an echo.MiddlewareFunc to be used with Echo.Use()
+func (e *EchoWrapper) Middleware() echo.MiddlewareFunc {
+	return func(next echo.HandlerFunc) echo.HandlerFunc {
+		return func(c echo.Context) error {
+			r := c.Request()
+			// get a new context with our trace from the request
+			ctx, span := common.StartSpanOrTraceFromHTTP(r)
+			defer span.Send()
+			// push the context with our trace and span on to the request
+			c.SetRequest(r.WithContext(ctx))
+
+			// get name of handler
+			handlerName := e.handlerName(c)
+			if handlerName == "" {
+				handlerName = "handler"
+			}
+			span.AddField("handler.name", handlerName)
+			span.AddField("name", handlerName)
+
+			// add field for each path param
+			for _, name := range c.ParamNames() {
+				span.AddField("app."+name, c.Param(name))
+			}
+
+			// add field for each query param
+			for name, values := range c.QueryParams() {
+				if len(values) == 1 {
+					span.AddField("app."+name, values[0])
+				} else {
+					span.AddField("app."+name, values)
+				}
+			}
+
+			// invoke next middleware in chain
+			err := next(c)
+
+			// add field for http response code
+			span.AddField("response.status_code", c.Response().Status)
+
+			return err
+		}
+	}
+}
+
+// Unfortunately the name of c.Handler() is an anonymous function
+// (https://github.com/labstack/echo/blob/master/echo.go#L487-L494).
+// This function will build a map of request paths to handle names during
+// the first request, thus providing quick lookup for every request thereafter.
+func (e *EchoWrapper) handlerName(c echo.Context) string {
```

Comment:
As per the comment, the name of the handler for a request isn't readily available so I'm building a map of request paths -> handler names for easy lookup.  Looking at the other example wrappers, the handler name seems important to add as a field to the honeycomb event?

Reply:
👍